File: java/agentbay/src/main/java/com/aliyun/agentbay/browser/Browser.java
package com.aliyun.agentbay.browser;

import java.util.HashMap;
import java.util.Map;

import com.aliyun.agentbay.Config;
import com.aliyun.agentbay.exception.BrowserException;
import com.aliyun.agentbay.model.OperationResult;
import com.aliyun.agentbay.service.BaseService;
import com.aliyun.agentbay.session.Session;
import com.aliyun.agentbay.util.ResponseUtil;
import com.aliyun.wuyingai20250506.models.InitBrowserRequest;
import com.aliyun.wuyingai20250506.models.InitBrowserResponse;
import com.aliyun.wuyingai20250506.models.InitBrowserResponseBody;
import com.fasterxml.jackson.databind.ObjectMapper;

/**
 * Browser provides browser-related operations for the session.
 */
public class Browser extends BaseService {
    private static final ObjectMapper objectMapper = new ObjectMapper();
    private static final String SERVER_BROWSER_USE = "wuying_browseruse";
    private static final String SERVER_CDP = "wuying_cdp_mcp_server";

    private String endpointUrl;
    private boolean initialized;
    private BrowserOption option;
    
    /**
     * The browser operator instance (recommended).
     */
    private BrowserOperator operator;
    /**
     * The browser agent instance (deprecated).
     */
    @Deprecated
    private BrowserAgent agent;
    
    private Integer endpointRouterPort;
    private volatile boolean agentDeprecationWarned = false;
    
    // Callback-related fields
    private BrowserCallback userCallback;
    private volatile boolean wsCallbackRegistered = false;
    private com.aliyun.agentbay._internal.WsClient.PushCallback internalWsCallback;

    public Browser(Session session) {
        super(session);
        this.initialized = false;
        
        this.operator = new BrowserOperator(session, this);
        this.agent = new BrowserAgent(session, this);
    }

    /**
     * Initialize the browser instance with the given options asynchronously.
     * Returns true if successful, false otherwise.
     * 
     * @param option Browser configuration options. If null, default options are used
     * @return true if initialization was successful, false otherwise
     */
    public boolean initialize(BrowserOption option) {
        if (isInitialized()) {
            return true;
        }

        try {
            InitBrowserRequest request = new InitBrowserRequest();
            request.setAuthorization("Bearer " + session.getApiKey());
            request.setSessionId(session.getSessionId());
            request.setPersistentPath(Config.BROWSER_DATA_PATH);

            Map<String, Object> browserOptionMap = option.toMap();

            // Set enableRecord based on session.enableBrowserReplay
            if (session.getEnableBrowserReplay() != null) {
                browserOptionMap.put("enableRecord", session.getEnableBrowserReplay());
            }

            // Convert BrowserOption to JSON string
            String browserOptionJson = objectMapper.writeValueAsString(browserOptionMap);
            request.setBrowserOption(browserOptionJson);

            InitBrowserResponse response = session.getAgentBay().getClient().initBrowser(request);
            String requestId = ResponseUtil.extractRequestId(response);

            if (response != null && response.getBody() != null && response.getBody().getData() != null) {
                // Extract port from response using the correct data type
                InitBrowserResponseBody.InitBrowserResponseBodyData data = response.getBody().getData();
                Integer port = data.getPort();
                if (port != null) {
                    this.endpointRouterPort = port;
                    this.initialized = true;
                    this.option = option;
                    return true;
                } else {
                }
            }
            return false;

        } catch (Exception e) {
            this.initialized = false;
            this.endpointUrl = null;
            this.option = null;
            return false;
        }
    }

    /**
     * Alias for initialize method.
     *
     * @param option Browser initialization options
     * @return true if successful, false otherwise
     */
    public boolean init(BrowserOption option) {
        return initialize(option);
    }

    /**
     * Destroy the browser instance manually.
     */
    public void destroy() {
        if (isInitialized()) {
            try {
                stopBrowser();
                // Reset browser state
                this.initialized = false;
                this.endpointUrl = null;
                this.option = null;
                this.endpointRouterPort = null;
            } catch (BrowserException e) {
            }
        }
    }

    /**
     * Takes a screenshot of the specified page with enhanced options and error handling.
     *
     * @param page The Playwright Page object to take a screenshot of. This is a required parameter.
     * @param fullPage Whether to capture the full scrollable page
     * @param options Additional screenshot options that will override defaults.
     *                Common options include:
     *                - type (ScreenshotType): Image type, either PNG or JPEG (default: PNG)
     *                - timeout (Double): Maximum time in milliseconds (default: 60000)
     *                - animations (String): How to handle animations (default: "disabled")
     *                - caret (String): How to handle the caret (default: "hide")
     *                - scale (String): Scale setting (default: "css")
     * @return Screenshot data as bytes
     * @throws BrowserException if browser is not initialized or page is null
     * @throws IllegalArgumentException if page is null
     */
    public byte[] screenshot(com.microsoft.playwright.Page page, boolean fullPage, Map<String, Object> options) throws BrowserException {
        if (!isInitialized()) {
            throw new BrowserException("Browser must be initialized before calling screenshot.");
        }
        if (page == null) {
            throw new IllegalArgumentException("Page cannot be null");
        }

        Map<String, Object> enhancedOptions = new HashMap<>();
        enhancedOptions.put("animations", "disabled");
        enhancedOptions.put("caret", "hide");
        enhancedOptions.put("scale", "css");
        enhancedOptions.put("timeout", options != null && options.containsKey("timeout") ? options.get("timeout") : 60000);
        enhancedOptions.put("fullPage", fullPage);
        enhancedOptions.put("type", options != null && options.containsKey("type") ? options.get("type") : com.microsoft.playwright.options.ScreenshotType.PNG);

        if (options != null) {
            enhancedOptions.putAll(options);
        }

        try {
            page.evaluate("window.scrollTo(0, document.body.scrollHeight)");
            page.waitForLoadState(com.microsoft.playwright.options.LoadState.DOMCONTENTLOADED,
                new com.microsoft.playwright.Page.WaitForLoadStateOptions().setTimeout(30000));

            scrollToLoadAllContent(page, 8, 1200);

            page.evaluate(
                "() => {\n" +
                "    document.querySelectorAll('img[data-src]').forEach(img => {\n" +
                "        if (!img.src && img.dataset.src) {\n" +
                "            img.src = img.dataset.src;\n" +
                "        }\n" +
                "    });\n" +
                "    document.querySelectorAll('[data-bg]').forEach(el => {\n" +
                "        if (!el.style.backgroundImage) {\n" +
                "            el.style.backgroundImage = `url(${el.dataset.bg})`;\n" +
                "        }\n" +
                "    });\n" +
                "}"
            );

            page.waitForTimeout(1500);

            int finalHeight = (int) page.evaluate("document.body.scrollHeight");
            page.setViewportSize(1920, Math.min(finalHeight, 10000));

            com.microsoft.playwright.Page.ScreenshotOptions screenshotOptions = new com.microsoft.playwright.Page.ScreenshotOptions();
            screenshotOptions.setFullPage((Boolean) enhancedOptions.get("fullPage"));
            screenshotOptions.setType((com.microsoft.playwright.options.ScreenshotType) enhancedOptions.get("type"));
            screenshotOptions.setTimeout((Double) enhancedOptions.getOrDefault("timeout", 60000.0));

            byte[] screenshotBytes = page.screenshot(screenshotOptions);
            return screenshotBytes;

        } catch (Exception e) {
            String errorMsg = "Failed to capture screenshot: " + e.getMessage();
            throw new BrowserException(errorMsg);
        }
    }

    /**
     * Scroll to load all content on the page.
     * 
     * @param page The Playwright Page object
     * @param maxScrolls Maximum number of scroll attempts (default: 8)
     * @param delayMs Delay between scrolls in milliseconds (default: 1200)
     */
    private void scrollToLoadAllContent(com.microsoft.playwright.Page page, int maxScrolls, int delayMs) {
        int lastHeight = 0;
        for (int i = 0; i < maxScrolls; i++) {
            page.evaluate("window.scrollTo(0, document.body.scrollHeight)");
            page.waitForTimeout(delayMs);
            int newHeight = (int) page.evaluate("Math.max(document.body.scrollHeight, document.documentElement.scrollHeight)");
            if (newHeight == lastHeight) {
                break;
            }
            lastHeight = newHeight;
        }
    }

    /**
     * Stop the browser instance (internal use only).
     * 
     * @throws BrowserException if browser is not initialized or stop operation fails
     */
    private void stopBrowser() throws BrowserException {
        if (!isInitialized()) {
            throw new BrowserException("Browser is not initialized. Cannot stop browser.");
        }

        try {
            OperationResult result = callMcpTool("stopChrome", new HashMap<>());
            if (!result.isSuccess()) {
            }
        } catch (Exception e) {
            throw new BrowserException("Failed to stop browser: " + e.getMessage());
        }
    }

    /**
     * Returns the endpoint URL if the browser is initialized, otherwise raises an exception.
     * When initialized, always fetches the latest CDP url from getCdpLink API.
     *
     * @return Browser endpoint URL
     * @throws BrowserException if browser is not initialized or endpoint URL cannot be retrieved
     * 
     */
    public String getEndpointUrl() throws BrowserException {
        if (!isInitialized()) {
            throw new BrowserException("Browser is not initialized. Cannot access endpoint URL.");
        }

        try {
            com.aliyun.wuyingai20250506.models.GetCdpLinkRequest request =
                new com.aliyun.wuyingai20250506.models.GetCdpLinkRequest();
            request.setAuthorization("Bearer " + session.getAgentBay().getApiKey());
            request.setSessionId(session.getSessionId());

            com.aliyun.wuyingai20250506.models.GetCdpLinkResponse response =
                session.getAgentBay().getClient().getCdpLink(request);

            if (response != null && response.getBody() != null &&
                response.getBody().getSuccess() && response.getBody().getData() != null) {
                this.endpointUrl = response.getBody().getData().getUrl();
            } else {
                String errorMsg = response != null && response.getBody() != null ?
                    response.getBody().getMessage() : "Unknown error";
                throw new BrowserException("Failed to get CDP link: " + errorMsg);
            }

            return this.endpointUrl;

        } catch (Exception e) {
            throw new BrowserException("Failed to get endpoint URL from session: " + e.getMessage());
        }
    }

    /**
     * Get the current BrowserOption used to initialize the browser.
     *
     * @return BrowserOption or null if not set
     * 
     */
    public BrowserOption getOption() {
        return option;
    }

    /**
     * Check if the browser is initialized.
     *
     * @return true if initialized, false otherwise
     * 
     */
    public boolean isInitialized() {
        return initialized;
    }

    /**
     * Get the browser operator for browser operations (recommended).
     * 
     * <p>The operator provides AI-powered browser automation capabilities including
     * navigation, screenshots, actions, observations, and data extraction.</p>
     * 
     * @return BrowserOperator instance
     */
    public BrowserOperator getOperator() {
        return operator;
    }

    /**
     * Get the browser agent for advanced browser operations.
     * 
     * <p><strong>⚠️ Deprecated</strong>: Use getOperator instead. This method will be removed in a future version.</p>
     * 
     * @return BrowserAgent instance
     * @deprecated Use getOperator instead
     */
    @Deprecated
    public BrowserAgent getAgent() {
        if (!agentDeprecationWarned) {
            synchronized (this) {
                if (!agentDeprecationWarned) {
                    System.err.println(
                        "[⚠️ DeprecationWarning] browser.agent is deprecated and will be removed in a future version. "
                        + "Please use browser.operator instead."
                    );
                    agentDeprecationWarned = true;
                }
            }
        }
        return agent;
    }

    /**
     * Get the endpoint router port.
     *
     * @return Port number or null if not set
     */
    public Integer getEndpointRouterPort() {
        return endpointRouterPort;
    }
    
    /**
     * Register a callback function to handle browser-related push notifications from sandbox.
     * 
     * <p>Example usage:</p>
     * <pre>{@code
     * BrowserCallback callback = (notifyMsg) -> {
     *     System.out.println("Type: " + notifyMsg.getType());
     *     System.out.println("Code: " + notifyMsg.getCode());
     *     System.out.println("Message: " + notifyMsg.getMessage());
     *     System.out.println("Action: " + notifyMsg.getAction());
     *     System.out.println("Extra params: " + notifyMsg.getExtraParams());
     * };
     * 
     * CreateResult createResult = agentBay.create();
     * Session session = createResult.getSession();
     * session.getBrowser().initialize(new BrowserOption());
     * boolean success = session.getBrowser().registerCallback(callback);
     * // ... do work ...
     * session.getBrowser().unregisterCallback();
     * session.delete();
     * }</pre>
     * 
     * @param callback Callback function that receives a BrowserNotifyMessage
     * @return true if the callback was successfully registered
     */
    public boolean registerCallback(BrowserCallback callback) {
        try {
            // Store user callback
            this.userCallback = callback;
            
            // Register internal callback to ws_client only once
            if (!wsCallbackRegistered) {
                com.aliyun.agentbay._internal.WsClient wsClient = session.getWsClient();
                if (wsClient == null) {
                    throw new BrowserException("WebSocket client is not available");
                }
                
                // Create internal callback wrapper
                internalWsCallback = (payload) -> {
                    try {
                        if (userCallback != null) {
                            BrowserNotifyMessage notifyMsg = BrowserNotifyMessage.fromMap(payload);
                            userCallback.onNotify(notifyMsg);
                        }
                    } catch (Exception e) {
                        System.err.println("Error in internal ws_callback: " + e.getMessage());
                    }
                };
                
                wsClient.registerCallback(SERVER_CDP, internalWsCallback);
                wsCallbackRegistered = true;
            }
            
            return true;
        } catch (Exception e) {
            System.err.println("Failed to register callback: " + e.getMessage());
            return false;
        }
    }
    
    /**
     * Unregister the previously registered callback function.
     * 
     * <p>Example usage:</p>
     * <pre>{@code
     * session.getBrowser().registerCallback(callback);
     * // ... do work ...
     * session.getBrowser().unregisterCallback();
     * }</pre>
     */
    public void unregisterCallback() {
        try {
            // Clear user callback
            userCallback = null;
            
            // Unregister from ws_client
            if (wsCallbackRegistered) {
                com.aliyun.agentbay._internal.WsClient wsClient = session.getWsClient();
                if (wsClient != null && internalWsCallback != null) {
                    wsClient.unregisterCallback(SERVER_CDP, internalWsCallback);
                    wsClient.close();
                }
                wsCallbackRegistered = false;
                internalWsCallback = null;
            }
        } catch (Exception e) {
            System.err.println("Failed to unregister callback: " + e.getMessage());
        }
    }
    
    /**
     * Send a notify message to sandbox through WebSocket.
     * 
     * <p>Example usage:</p>
     * <pre>{@code
     * BrowserNotifyMessage notifyMsg = new BrowserNotifyMessage(
     *     "call-for-user",
     *     1,
     *     199,
     *     "user handle done",
     *     "takeoverdone",
     *     new HashMap<>()
     * );
     * boolean success = session.getBrowser().sendNotifyMessage(notifyMsg);
     * }</pre>
     * 
     * @param notifyMessage The notification message to send
     * @return true if the message was successfully sent, false otherwise
     */
    public boolean sendNotifyMessage(BrowserNotifyMessage notifyMessage) {
        try {
            com.aliyun.agentbay._internal.WsClient wsClient = session.getWsClient();
            if (wsClient == null) {
                throw new BrowserException("WebSocket client is not available");
            }
            
            // Send notify message through ws_client
            wsClient.sendMessage(SERVER_CDP, notifyMessage.toMap());
            return true;
        } catch (Exception e) {
            System.err.println("Failed to send notify message: " + e.getMessage());
            return false;
        }
    }
    
    /**
     * Send a takeoverdone notify message to sandbox.
     * 
     * <p>Example usage:</p>
     * <pre>{@code
     * BrowserCallback callback = (notifyMsg) -> {
     *     if ("takeover".equals(notifyMsg.getAction())) {
     *         int takeoverNotifyId = notifyMsg.getId();
     *         // ... do work in other thread...
     *         session.getBrowser().sendTakeoverDone(takeoverNotifyId);
     *     }
     * };
     * 
     * CreateResult createResult = agentBay.create();
     * Session session = createResult.getSession();
     * session.getBrowser().initialize(new BrowserOption());
     * session.getBrowser().registerCallback(callback);
     * // ... do work ...
     * session.getBrowser().unregisterCallback();
     * session.delete();
     * }</pre>
     * 
     * @param notifyId The notification ID associated with the takeover request message
     * @return true if the takeoverdone notify message was successfully sent, false otherwise
     */
    public boolean sendTakeoverDone(int notifyId) {
        try {
            // Get ws_client
            com.aliyun.agentbay._internal.WsClient wsClient = session.getWsClient();
            if (wsClient == null) {
                throw new BrowserException("WebSocket client is not available");
            }
            
            // Build takeoverdone notify message
            BrowserNotifyMessage notifyMessage = new BrowserNotifyMessage(
                "call-for-user",
                notifyId,
                199,
                "user handle done",
                "takeoverdone",
                new HashMap<>()
            );
            
            // Send message through ws_client
            wsClient.sendMessage(SERVER_CDP, notifyMessage.toMap());
            return true;
        } catch (Exception e) {
            System.err.println("Failed to send browser notify message: " + e.getMessage());
            return false;
        }
    }
}

File: python/agentbay/_async/browser.py
import asyncio
import base64
import os
import time
import typing
from typing import TYPE_CHECKING, Literal, Optional, Union

from .._common.config import _BROWSER_DATA_PATH
from .._common.exceptions import BrowserError
from .._common.logger import _log_api_response_with_details, get_logger
from ..api.models import InitBrowserRequest
from .base_service import AsyncBaseService
from .browser_agent import AsyncBrowserAgent

# Initialize logger for this module
_logger = get_logger("browser")

if TYPE_CHECKING:
    from .._common.models import FingerprintFormat
    from .._common.models import BrowserOption
    from .session import AsyncSession

class AsyncBrowser(AsyncBaseService):
    """
    Browser provides browser-related operations for the session.
    """

    def __init__(self, session: "AsyncSession"):
        self.session = session
        self._endpoint_url = None
        self._initialized = False
        self._option = None
        self.agent = AsyncBrowserAgent(self.session, self)
        self.endpoint_router_port = None

    async def initialize(self, option: Optional["BrowserOption"] = None) -> bool:
        """
        Initialize the browser instance with the given options asynchronously.
        Returns True if successful, False otherwise.

        Args:
            option (BrowserOption, optional): Browser configuration options. If None, default options are used.

        Returns:
            bool: True if initialization was successful, False otherwise.

        Example:
            ```python
            create_result = await agent_bay.create()
            session = create_result.session
            # Use default options
            await session.browser.initialize()
            # Or with specific options
            browser_option = BrowserOption(use_stealth=True)
            await session.browser.initialize(browser_option)
            await session.delete()
            ```
        """
        if self.is_initialized():
            return True

        if option is None:
            option = BrowserOption()

        try:
            browser_option_dict = option._to_map()

            # Enable record if session.enableBrowserReplay is True
            if (
                hasattr(self.session, "enableBrowserReplay")
                and self.session.enableBrowserReplay
            ):
                browser_option_dict["enableRecord"] = True

            request = InitBrowserRequest(
                authorization=f"Bearer {self.session._get_api_key()}",
                session_id=self.session._get_session_id(),
                persistent_path=_BROWSER_DATA_PATH,
                browser_option=browser_option_dict,
            )
            # Use async client method
            response = await self.session._get_client().init_browser_async(request)
            _logger.debug(f"Response from init_browser: {response}")
            response_map = response.to_map()
            body = response_map.get("body", {})
            data = body.get("Data", {})
            success = data.get("Port") is not None
            if success:
                self.endpoint_router_port = data.get("Port")
                self._initialized = True
                self._option = option
                _log_api_response_with_details(
                    api_name="InitBrowser (async)",
                    success=True,
                    key_fields={
                        "port": data.get("Port"),
                        "status": "successfully initialized",
                    },
                )
                _logger.info("Browser instance successfully initialized")
            return success
        except Exception as e:
            _logger.exception(
                f"❌ Failed to initialize browser instance asynchronously"
            )
            self._initialized = False
            self._endpoint_url = None
            self._option = None
            return False

    async def init(self, option: Optional["BrowserOption"] = None) -> bool:
        """
        Alias for initialize.
        """
        return await self.initialize(option)

    async def destroy(self):
        """
        Destroy the browser instance manually.
        """
        await self._stop_browser()

    async def screenshot(self, page, full_page: bool = False, **options) -> bytes:
        """
        Takes a screenshot of the specified page with enhanced options and error handling.

        Args:
            page (Page): The Playwright Page object to take a screenshot of. This is a required parameter.
            full_page (bool): Whether to capture the full scrollable page. Defaults to False.
            **options: Additional screenshot options that will override defaults.
                      Common options include:
                      - type (str): Image type, either 'png' or 'jpeg' (default: 'png')
                      - quality (int): Quality of the image, between 0-100 (jpeg only)
                      - timeout (int): Maximum time in milliseconds (default: 60000)
                      - animations (str): How to handle animations (default: 'disabled')
                      - caret (str): How to handle the caret (default: 'hide')
                      - scale (str): Scale setting (default: 'css')

        Returns:
            bytes: Screenshot data as bytes.

        Raises:
            BrowserError: If browser is not initialized.
            RuntimeError: If screenshot capture fails.
        """
        # Check if browser is initialized
        if not self.is_initialized():
            raise BrowserError("Browser must be initialized before calling screenshot.")
        if page is None:
            raise ValueError("Page cannot be None")
        # Set default enhanced options
        enhanced_options = {
            "animations": "disabled",
            "caret": "hide",
            "scale": "css",
            "timeout": options.get("timeout", 60000),
            "full_page": full_page,  # Use the function parameter, not options
            "type": options.get("type", "png"),
        }

        # Update with user-provided options (but full_page is already set from function parameter)
        enhanced_options.update(options)

        try:
            # Wait for page to load
            # await page.wait_for_load_state("networkidle")
            await page.evaluate("window.scrollTo(0, document.body.scrollHeight)")
            await page.wait_for_load_state("domcontentloaded", timeout=30000)
            # Scroll to load all content (especially for lazy-loaded elements)
            await self._scroll_to_load_all_content_async(page)

            # Ensure images with data-src attributes are loaded
            await page.evaluate(
                """
                () => {
                    document.querySelectorAll('img[data-src]').forEach(img => {
                        if (!img.src && img.dataset.src) {
                            img.src = img.dataset.src;
                        }
                    });
                    // Also handle background-image[data-bg]
                    document.querySelectorAll('[data-bg]').forEach(el => {
                        if (!el.style.backgroundImage) {
                            el.style.backgroundImage = `url(${el.dataset.bg})`;
                        }
                    });
                }
            """
            )

            # Wait a bit for images to load
            await page.wait_for_timeout(1500)
            final_height = await page.evaluate("document.body.scrollHeight")
            await page.set_viewport_size(
                {"width": 1920, "height": min(final_height, 10000)}
            )

            # Take the screenshot
            screenshot_bytes = await page.screenshot(**enhanced_options)
            _logger.info("Screenshot captured successfully.")
            return screenshot_bytes

        except Exception as e:
            # Convert exception to string safely to avoid comparison issues
            try:
                error_str = str(e)
            except:
                error_str = "Unknown error occurred"
            error_msg = f"Failed to capture screenshot: {error_str}"
            _logger.error(error_msg)
            raise RuntimeError(error_msg) from e

    async def _scroll_to_load_all_content_async(
        self, page, max_scrolls: int = 8, delay_ms: int = 1200
    ):
        """Async version of _scroll_to_load_all_content."""
        last_height = 0
        for _ in range(max_scrolls):
            await page.evaluate("window.scrollTo(0, document.body.scrollHeight)")
            await page.wait_for_timeout(delay_ms)
            new_height = await page.evaluate(
                "Math.max(document.body.scrollHeight, document.documentElement.scrollHeight)"
            )
            if new_height == last_height:
                break
            last_height = new_height

    async def _stop_browser(self):
        """
        Stop the browser instance, internal use only.
        """
        if self.is_initialized():
            await self.session.call_mcp_tool("stopChrome", {})
        else:
            raise BrowserError("Browser is not initialized. Cannot stop browser.")

    async def get_endpoint_url(self) -> str:
        """
        Returns the endpoint URL if the browser is initialized, otherwise raises an exception.
        When initialized, always fetches the latest CDP url from session.get_link().

        Returns:
            str: The browser CDP endpoint URL.

        Raises:
            BrowserError: If browser is not initialized or endpoint URL cannot be retrieved.

        Example:
            ```python
            create_result = await agent_bay.create()
            session = create_result.session
            browser_option = BrowserOption()
            await session.browser.initialize(browser_option)
            endpoint_url = await session.browser.get_endpoint_url()
            print(f"CDP Endpoint: {endpoint_url}")
            await session.delete()
            ```
        """
        if not self.is_initialized():
            raise BrowserError(
                "Browser is not initialized. Cannot access endpoint URL."
            )
        try:
            if self.session.is_vpc:
                _logger.debug(
                    f"VPC mode, endpoint_router_port: {self.endpoint_router_port}"
                )
                self._endpoint_url = f"ws://{self.session.network_interface_ip}:{self.endpoint_router_port}"
            else:
                from ..api.models import GetCdpLinkRequest

                request = GetCdpLinkRequest(
                    authorization=f"Bearer {self.session.agent_bay.api_key}",
                    session_id=self.session.session_id,
                )
                # Async call
                response = await self.session.agent_bay.client.get_cdp_link_async(
                    request
                )
                if response.body and response.body.success and response.body.data:
                    self._endpoint_url = response.body.data.url
                else:
                    error_msg = (
                        response.body.message if response.body else "Unknown error"
                    )
                    raise BrowserError(f"Failed to get CDP link: {error_msg}")
            return self._endpoint_url
        except Exception as e:
            raise BrowserError(f"Failed to get endpoint URL from session: {e}")

    def get_option(self) -> Optional["BrowserOption"]:
        """
        Returns the current BrowserOption used to initialize the browser, or None if not set.

        Returns:
            Optional[BrowserOption]: The browser options if initialized, None otherwise.

        Example:
            ```python
            create_result = await agent_bay.create()
            session = create_result.session
            browser_option = BrowserOption(use_stealth=True)
            await session.browser.initialize(browser_option)
            current_options = session.browser.get_option()
            print(f"Stealth mode: {current_options.use_stealth}")
            await session.delete()
            ```
        """
        return self._option

    def is_initialized(self) -> bool:
        """
        Returns True if the browser was initialized, False otherwise.

        Returns:
            bool: True if browser is initialized, False otherwise.

        Example:
            ```python
            create_result = await agent_bay.create()
            session = create_result.session
            print(f"Initialized: {session.browser.is_initialized()}")
            browser_option = BrowserOption()
            await session.browser.initialize(browser_option)
            print(f"Initialized: {session.browser.is_initialized()}")
            await session.delete()
            ```
        """
        return self._initialized
